feat(opentelemetry-instrumentation-fetch): added a requestHook option

patricklafrance opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/main/guides/contributor#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes https://github.com/open-telemetry/opentelemetry-js/issues/5084

Add a `requestHook` option to the `@opentelemetry/instrumentation-fetch` package.

## Short description of the changes

Add a `requestHook` option to the `@opentelemetry/instrumentation-fetch` package.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update (The README file of the package)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

Added the following unit tests to the `@opentelemetry/instrumentation-fetch` package:

```
`requestHook` option
    can apply attributes to the span
    can modify headers when called with a string URL
    can modify headers when called with a `Request` object
    can modify headers when called with a `RequestInit` object
```

The tests can be executed by following those instructions:

```bash
cd experimental/packages/opentelemetry-instrumentation-fetch
npm run test:browser
````

## Checklist:

- [x] Followed the style guidelines of this project
- [x] Unit tests have been added
- [x] Documentation has been updated

chancancode commented:
@patricklafrance hey sorry for causing the conflicts in the tests, they have been completely rewritten in #5282.

Instead of having one big shared fake-fetch function that we keep adding to to support all the various things we want to test, now each group of test have a smaller and more focused setup block specific to their needs. Hopefully you'd ultimately find that easier to work with, despite the trouble.

Without looking into this too deeply, I think the tests for `applyCustomAttributesOnSpan` (L1211-1368) may be a good starting point. https://github.com/open-telemetry/opentelemetry-js/pull/5282/commits/72089a5ac8bba399e0db0c88078b2f781debb7d6 is the isolated commit where I ported those tests.

Let me know if you need help working through that.

patricklafrance commented:
Hey @chancancode I rewrote the tests using the new structure. Because I need to assert something on the "tempered" fetch request I added back a "localized" `fake-fetch` function. I hope it's fine.

Could you by any chance review this PR and consider merging it? That would be greatly appreciated 🙏🏻

chancancode commented on experimental/packages/opentelemetry-instrumentation-fetch/test/fetch.test.ts:
```diff
@@ -1367,6 +1367,89 @@ describe('fetch', () => {
+        const rootSpan = await trace(async () => {
+          await callback();
+        }, config);
+
+        fetchStub.reset();
+
+        assert.strictEqual(exportedSpans.length, 1);
+
+        return { rootSpan, request: request as RequestInit };
+      };
+
+      it('applies attributes to the span', async () => {
+        await tracedFetch({
+          config: {
+            requestHook: span => {
+              span.setAttribute('custom.foo', 'bar');
+            },
+          },
+        });
+
+        const span: tracing.ReadableSpan = exportedSpans[0];
+        assert.strictEqual(span.attributes['custom.foo'], 'bar');
+      });
+
+      it('applies headers to the request', async () => {
+        const { request: traceRequest } = await tracedFetch({
+          config: {
+            requestHook: (span, request) => {
+              (request.headers as Record<string, string>)['traceparent'] =
+                'custom-value';
+            },
+          },
+        });
+
+        assert.strictEqual(
+          (traceRequest.headers as Record<string, string>)['traceparent'],
+          'custom-value'
+        );
+      });
+    });
```
I am not a maintainer so I can't merge your PR and you should take my suggestions with a grain of salt.

But I don't think it should be necessary to stub fetch in your tests – that's the main issue with the previous tests and what we tried to move away from. It's too easy to mock the browser feature in incorrect/incomplete ways and end up testing behaviors that doesn't exist in the real world, or mask breakages that can only be discovered in production.

Something like this should be sufficient to ensure the feature is working correctly, as I understood it. You should review this carefully to make sure that's the case. I'm also just typing this out in the browser, there may be syntax errors and you may have to make adjustments for TypeScript/eslint/prettier.

```suggestion
    describe('`requestHook` config', () => {
      const tracedFetch = async ({
        handlers = [
          msw.http.get('/api/echo-headers.json', ({ request }) => {
            return msw.HttpResponse.json({
              request: {
                headers: Object.fromEntries(request.headers),
              },
            });
          }),
        ],
        callback = () => fetch('/api/echo-headers.json'),
        config,
      }: {
        handlers?: msw.RequestHandler[];
        callback?: () => Promise<Response>;
        config: FetchInstrumentationConfig &
          Required<Pick<FetchInstrumentationConfig, 'requestHook'>>;
      }): Promise<{ rootSpan: api.Span; request: RequestInit }> => {
        let response: Response | undefined;

        await startWorker(...handlers);

        const rootSpan = await trace(async () => {
          response = await callback();
        }, config);

        assert.ok(response instanceof Response);
        assert.strictEqual(exportedSpans.length, 1);

        return { rootSpan, response };
      };

      it('can apply attributes to the span', async () => {
        await tracedFetch({
          config: {
            requestHook: span => {
              span.setAttribute('custom.foo', 'bar');
            },
          },
        });

        const span: tracing.ReadableSpan = exportedSpans[0];
        assert.strictEqual(span.attributes['custom.foo'], 'bar');
      });

      it('can remove propagation headers', async () => {
        const { response } = await tracedFetch({
          config: {
            requestHook: (span, request) => {
              assert.ok(
                request !== null && typeof request === 'object' && !(request instanceof Request),
                '`requestHook` should get a `RequestInit` object when no options are passed to `fetch()`'
              );

              assert.ok(
                request.headers !== null && typeof request.headers === 'object',
                '`requestHook` should get the `headers` object generated by the instrumentation'
              );

              assert.ok(
                X_B3_TRACE_ID in request.headers,
                `trace header '${X_B3_TRACE_ID}' not set`
              );

              assert.ok(
                X_B3_SPAN_ID in request.headers,
                `trace header '${X_B3_SPAN_ID}' not set`
              );

              assert.ok(
                X_B3_SAMPLED in request.headers,
                `trace header '${X_B3_SAMPLED}' not set`
              );

              delete request.headers[X_B3_TRACE_ID];
              delete request.headers[X_B3_SPAN_ID];
              delete request.headers[X_B3_SAMPLED];
            },
          },
        });

        await assertNoPropagationHeaders(response);
      });
      
      it('can modify headers when called with a string URL', async () => {
        const { response } = await tracedFetch({
          config: {
            requestHook: (span, request) => {
              assert.ok(
                request !== null && typeof request === 'object' && !(request instanceof Request),
                '`requestHook` should get a `RequestInit` object when no options are passed to `fetch()`'
              );
              request.headers = { 'custom-foo': 'foo' };
            },
          },
        });
        
        const headers = await assertPropagationHeaders(response);

        assert.strictEqual(
          headers['custom-foo'],
          'foo',
          'header set from requestHook should be sent',
        );
      });

      it('can modify headers when called with a `Request` object', async () => {
        const { response } = await tracedFetch({
          config: {
            requestHook: (span, request) => {
              assert.ok(
                request instanceof Request,
                '`requestHook` should get the `Request` object passed to `fetch()`'
              );

              request.headers.set('custom-foo', 'foo');
            },
          },
          callback: () =>
            fetch(
              new Request('/api/echo-headers.json', {
                headers: new Headers({ 'custom-bar: 'bar' }),
              })
        });

        const headers = await assertPropagationHeaders(response);

        assert.strictEqual(
          headers['custom-foo'],
          'foo',
          'header set from requestHook should be sent',
        );
        assert.strictEqual(
          headers['custom-bar'],
          'foo',
          'header set from fetch() should be sent',
        );
      });

      it('can modify headers when called with a `RequestInit` object', async () => {
        const { response } = await tracedFetch({
          config: {
            requestHook: (span, request) => {
              assert.ok(
                request !== null && typeof request === 'object' && !(request instanceof Request),
                '`requestHook` should get the `RequestInit` object passed to `fetch()`'
              );
              
              assert.ok(
                request.headers !== null && typeof request.headers === 'object',
                '`requestHook` should get the `headers` object passed to `fetch()`'
              );

              (request.headers as Record<string, string>)['custom-foo'] = 'foo';
            },
          },
          callback: () =>
            fetch('/api/echo-headers.json', {
              headers: { 'custom-bar: 'bar' },
            })
        });

        const headers = await assertPropagationHeaders(response);

        assert.strictEqual(
          headers['custom-foo'],
          'foo',
          'header set from requestHook should be sent',
        );
        assert.strictEqual(
          headers['custom-bar'],
          'foo',
          'header set from fetch() should be sent',
        );
      });
    });
```

The idea is that we defined a mock api endpoint ("handler" in msw) that simply slurps up any request headers you sent and echo it back as JSON in the response body, so you can make a normal request, have it fully flow through all the real `fetch()` function and the network layers in the browser. That way you are testing the real, end-to-end behavior, rather than narrowing mocking just the one thing you expect to happen.

Hope that makes sense.

patricklafrance replied:
Smart, thank you, I'll try what you are suggesting next week 🙏🏻

chancancode replied:
@patricklafrance glad that was helpful! not sure if you saw but in that gist I added a specific test for overriding (removing) the propagation headers. If that’s the specific use case that motivated you to add this feature, you may want to make sure it’s tested so it doesn’t regress

chancancode commented:
As an aside, personally I dislike how we are try to preserve the inputs to the `fetch()` call to the fullest extent.

It makes the internally instrument code more complicated than it should be, as every place that wants to make modifications to the options has to consider all of those possibilities. But as demonstrated in the tests, it also pushes that complexity down to the consumers, as their hooks also have to be prepared to work with any number of those possibilities. Even if your application's coding style enforces a consistent call-style, one day, you may add a dependency or script that calls `fetch()` with a different style and your hooks will break if you are not prepared for it.

If it's up to me, I'd refactor the patched fetch function to always normalize the options into a `Request` object upfront, and then the rest of the code, including the hooks, can assume that.

In any case that's probably out of scope for this PR, so it's not really your problem, and it's not very actionable for you. But this new feature does introduce another spot where this currently not ideal (IMO) semantics leaks out to the consumers that we may have to break if we want to do what I proposed, and I think this is a good opportunity to bring that up for consideration.

patricklafrance commented:
> As an aside, personally I dislike how we are try to preserve the inputs to the `fetch()` call to the fullest extent.
> 
> It makes the internally instrument code more complicated than it should be, as every place that wants to make modifications to the options has to consider all of those possibilities. But as demonstrated in the tests, it also pushes that complexity down to the consumers, as their hooks also have to be prepared to work with any number of those possibilities. Even if your application's coding style enforces a consistent call-style, one day, you may add a dependency or script that calls `fetch()` with a different style and your hooks will break if you are not prepared for it.
> 
> If it's up to me, I'd refactor the patched fetch function to always normalize the options into a `Request` object upfront, and then the rest of the code, including the hooks, can assume that.
> 
> In any case that's probably out of scope for this PR, so it's not really your problem, and it's not very actionable for you. But this new feature does introduce another spot where this currently not ideal (IMO) semantics leaks out to the consumers that we may have to break if we want to do what I proposed, and I think this is a good opportunity to bring that up for consideration.

I get it. This is not my preferred solution as well. We are Honeycomb customers and are trying to integrate Honeycomb tracing to our frontend framework using OTel. Unfortunately, we seem to have hit a limitation of the current API that many others stumble upon as well according to [this issue](https://github.com/open-telemetry/opentelemetry-js/issues/3558). At the moment, adding this hook seems to be the easiest way to achieve our goal. We've been staled since last November.

patricklafrance commented:
Thanks @chancancode I updated the tests following your suggestions, way better! 🙏🏻

patricklafrance commented on experimental/packages/opentelemetry-instrumentation-fetch/README.md:
```diff
@@ -68,9 +68,10 @@ Fetch instrumentation plugin has few options available to choose from. You can s
-| [`ignoreNetworkEvents`](https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/opentelemetry-instrumentation-fetch/src/fetch.ts#L77)         | `boolean`                     | Disable network events being added as span events (network events are added by default) |
-| [`measureRequestSize`](https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/opentelemetry-instrumentation-fetch/src/fetch.ts#L79)          | `boolean`                     | Measure outgoing request length (outgoing request length is not measured by default)    |
+| [`applyCustomAttributesOnSpan`](https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/opentelemetry-instrumentation-fetch/src/fetch.ts#L83) | `FetchCustomAttributeFunction` | Function for adding custom attributes                                                   |
+| [`requestHook`](https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/opentelemetry-instrumentation-fetch/src/fetch.ts#L85) | `FetchRequestHookFunction` | Function for adding custom attributes or headers before the request is handled                                                  |
```
Hey @JamieDanielson, still not perfect but I improved the alignment. This one is more complicated to aligned than the Semantic Conventions table because it includes super long links.

https://github.com/open-telemetry/opentelemetry-js/pull/5380/files#diff-e88ad968f43d1a12f14e88b4ea3b260db73a6da37aa6bf436bf83a45ae243f96R69

JamieDanielson replied:
Ah it's not a big deal at all. I've noticed some markdown files are consistent with styles and some aren't. My editor keeps auto-formatting so I may just need to go through in a sweep and update some places in the codebase 😅  This one only stood out because it was new but it renders the same.

patricklafrance commented on experimental/packages/opentelemetry-instrumentation-fetch/test/fetch.test.ts:
```diff
@@ -1367,6 +1367,150 @@ describe('fetch', () => {
+                '`requestHook` should get the `Request` object passed to `fetch()`'
+              );
+
+              request.headers.set('custom-foo', 'foo');
```
I added a few assertions suggesting that the `User-Agent` header cannot be overridden:

- https://github.com/open-telemetry/opentelemetry-js/pull/5380/files#diff-e410f489fdbd91ab6efe08b360062083668775846340c14272b5c93e6c955f68R1438
- https://github.com/open-telemetry/opentelemetry-js/pull/5380/files#diff-e88ad968f43d1a12f14e88b4ea3b260db73a6da37aa6bf436bf83a45ae243f96R69
- https://github.com/open-telemetry/opentelemetry-js/pull/5380/files#diff-e410f489fdbd91ab6efe08b360062083668775846340c14272b5c93e6c955f68R1524
-

chancancode replied:
I’m not sure that it is very relevant here. We expose the raw Request/RequestInit object, so you can do whatever the browser would let you do.

I don’t know off the top of my head whether the browser has any restrictions on setting that header, but I dont think we specifically introduced any additional roadblocks/restrictions in our code right? A casual search on SO also seems to suggest the answer may be browser-specific, so testing for that here may be overly specific/brittle IMO. 

Also may want to check the casing is correct since the browser normalize the built-in headers.

JamieDanielson replied:
I brought up the User Agent scenario specifically because it is a semantic convention and also because I recall differences between backend and frontend permissions with APIs. I remember a while back when trying to update the signal specific headers, I ran into the issue of not being able to alter User Agent and was curious if this was still the case. I think it's useful to have a test asserting behavior of this special header, because if it does break that's an indication of potential unexpected breakage in telemetry as well.

JamieDanielson replied:
Hmm... I can see where `assert.notEqual(...['User-Agent']...)` could be brittle though. I like the idea of asserting the User Agent cannot be overridden, but wonder if a) we should move it into its own separate test block to call it out more explicitly and b) if we can think of another way to test it that isn't notEqual. For example can we get the user agent being defined in the test setup and assert that it has not changed?

patricklafrance replied:
The reason I added it the existing tests is because there is actually “3 syntaxes” to override the header.

I can add 3 additional tests if you prefer, let me know.

chancancode replied:
>  was curious if this was still the case

I was able to test that real quick using https://postman-echo.com/get and this snippet:

```js
(async () => {
  let response = await fetch('https://postman-echo.com/get', { headers: { 'user-agent': 'foo'} });
  let json = await response.json();
  console.log(json);
})()
```

The result is indeed browser-specific. I believe the current state is that the spec allows it to be set (or rather, no longer forbid it). In Chrome, that UA header passed into `fetch()` doesn't do anything, but in Safari (and allegedly Firefox), it does work, which I was able to confirm locally.

So if we do add tests for it, it would be testing Chrome-specific behavior, which is fine I suppose, but IMO we should at least make a note of that in the code comment.

After #5282, the tests can actually run on Firefox now by adding the `karma-firefox-launcher` package/plugin, and I was going to propose that we start doing that (not in this PR). If we eventually proceed with that, it would be useful to know that this test is not expected to pass everywhere.

(And also it does look like the test currently maybe a problematic due to casing, probably want to assert that it equals `navigator.userAgent` or something.)

JamieDanielson replied:
I was just about messing around with it and confirmed the assertion isn't actually what we are expecting in some places. For stringURL and RequestInit Object tests, it is making assertions against an object for `request.headers` that looks like `Object{accept: '*/*', custom-foo: 'foo', useragent: 'custom-ua'}`. By changing to assert against the actual user agent header, we can see how it behaves. 


```js
const DEFAULT_USER_AGENT = navigator.userAgent;
// 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/132.0.0.0 Safari/537.36'
...
// this passes but not for the reasons we want
assert.notEqual(
  request.headers['User-Agent'],
  'custom-ua',
  'User-Agent shoud not be overrided'
);
// this fails in StringURL and RequestInit Object tests, but passes when called with Request Object
assert.strictEqual(
  request.headers['User-Agent'],
  DEFAULT_USER_AGENT,
  'User-Agent should not be overridden'
);
```

JamieDanielson replied:
So my thought in bringing all of this up was just to be explicit about behavior with a "special" header, since we're adding a feature that specifically references the ability to override headers. What do y'all think about adding the User Agent test to the Request Object, using the `navigator.userAgent` as a constant as shown above, and maybe just omitting it from the other tests?

JamieDanielson replied:
Interesting to note the behavior is different in Firefox 🤔  In that case, it's definitely worth the comment for now... or we can remove it entirely. It is certainly possible I have the User Agent item stuck in my brain and it's not really useful right now.

chancancode replied:
Good catch on the assertions!

Yea, it's a strange situation. My quick read on the history is that, at one point, the spec explicitly forbid changing that header. After some lengthy discussion, that ban was dropped, and most browsers followed through, except Chrome (and presumably Edge?) is still holding out due to some unspecified concerns.

I think you also made a good point that we currently set the client UA semconv attribute, and we implicitly assume that it can't be overridden, since we hardcoded the attribute to `navigator.userAgent`, which.. now that we discussed this a bunch, is arguably a bug in the browsers that does allow you to override it.

So I think a dedicated test for the currently expected UA header behavior as you suggested, with a comment to investigate further for other browsers, is probably a good idea. We can then punt the work to add Firefox to the roster and adjust the behavior and test when someone gets the time to look into that whole thing?

patricklafrance replied:
@JamieDanielson Would you mind merging this PR without adding additional tests related to the user agent? There seems to be many quirks with the user agent header that are kind of related to this PR but also seems to be larger than this PR.

Let me know 🙏🏻

JamieDanielson replied:
Yes I think we should remove all of the user agent tests that were added, as we realized it's not testing what we thought it was testing. Sorry for adding extra work to your plate! I guess we've learned some things along the way 😅  and yes the user agent behavior should probably be investigated and sorted out but is out of scope for this PR.

patricklafrance replied:
No worries @JamieDanielson. I cleaned up the tests, should be good to merge now :)

JamieDanielson reviewed (approved):
This looks good, thanks for working on it. Main thing right now is that we are finishing up SDK2.0 and have a feature freeze in place for just over a week longer. I will check in with the other maintainers on timing. 

Also if anyone who works more closely with web has thoughts on this please chime in!

pichlermarc commented:
Merging this as it looks like the only question left are on release timing.
